### DISCUSSION TITLE
Is there a way to add our own Tweens (as part of original library)

### DISCUSSION THREAD

**Discussion by AnthonyDemanueleFSG** [KyryloKuzyk/PrimeTween#122] [General] [answered]
For example, I added a custom tween as follows:

```
public static Tween UpdateText(TMP_Text text, int startValue, int endValue, string formattedString, int arg1, TweenSettings tweenSettings)
{
     return Tween.Custom(text, startValue, endValue, tweenSettings.duration, (target, val) => target.SetText(formattedString, val, arg1), tweenSettings.ease);
 }
```
 
but to call this I need to use something like:
```
.Group(Tween.UIFillAmount(SampleFillImage, 0.5f, TweenSettings))
.Group(UpdateText(SamlpeText, 100,200, "{0:0}", 0, TweenSettings))
```

What I would like is to be able to add Tweens as part of the same namespace so you'd call the new Tween something like this:
```
.Group(Tween.UIFillAmount(SampleFillImage, 0.5f, TweenSettings))
.Group(Tween.UpdateText(SamlpeText, 100,200, "{0:0}", 0, TweenSettings))
```

**Comment by KyryloKuzyk** [comment, marked as answer]
Hey Anthony,

There is no simple way to add a built-in animation. You can, of course, embed the library by copying its source code to the Packages folder, but that's not a good long-term solution.

The reason why this animation can't be built-in is because it would break PrimeTween's zero-allocation design. Your example allocates garbage because the delegate captures formattedString and arg1 arguments into a closure. Here is an example of a possible solution:

```csharp
[SerializeField] Image SampleFillImage;
[SerializeField] TextAnimationData textAnimationData;
[SerializeField] TweenSettings TweenSettings;

public Sequence Animate() {
    return Sequence.Create()
        .Group(Tween.UIFillAmount(SampleFillImage, 0.5f, TweenSettings))
        .Group(textAnimationData.Animate(100, 200, TweenSettings));
}

[Serializable]
public class TextAnimationData {
    public TMP_Text text;
    public string formattedString;
    public int arg1;

    public Tween Animate(float startValue, float endValue, TweenSettings tweenSettings)
    {
        return Tween.Custom(this, startValue, endValue, tweenSettings, (t, v) => t.text.SetText(t.formattedString, v, t.arg1));
    }
}
```

**Reply by AnthonyDemanueleFSG** [reply]
Understood - thank you!

**Reply by krisrok** [reply]
There _is_ a way using an assembly reference, which is an underrated quirky unity feature. Just create an assembly reference pointing to `PrimeTween.Runtime` and put your own extensions into a file next to the .asmref copying the original's structure like so:
```cs
namespace PrimeTween
{
    public partial struct Tween
    {
        public static Tween UpdateText(TMP_Text text, int startValue, int endValue, string formattedString, int arg1, TweenSettings tweenSettings)
        {
            return Tween.Custom(text, startValue, endValue, tweenSettings.duration, (target, val) => target.SetText(formattedString, val, arg1), tweenSettings.ease);
        }
    }
}
```

That said, the above technique has some niche applications e.g. when it comes to avoiding messing with reflection.

But in your case, I'd just drop your extension into a custom class, like `MyTween` or `TweenExtensions` and be done with it :)
